# Step 3: Combining the Datasets

Let's import our packages

In [1]:
# imports
import pandas as pd
import numpy as np

**Begin by merging the data we got from the APIs with the population data for each state**

First, we import the files

In [2]:
# df_quality = pd.read_csv('Article_data_with_RevIDs_Scores.csv')
df_quality = pd.read_csv('Article_data_with_RevIDs_Scores_Full_Valid.csv')
df_quality

revision_id           page_title    state article_quality  \
0           104730   Abbeville, Alabama  Alabama            Stub   
1           104761  Adamsville, Alabama  Alabama            Stub   
2           105188     Addison, Alabama  Alabama            Stub   
3           104726       Akron, Alabama  Alabama            Stub   
4           105109   Alabaster, Alabama  Alabama            Stub   
...            ...                  ...      ...             ...   
20152       140221   Wamsutter, Wyoming  Wyoming           Start   
20153       140185   Wheatland, Wyoming  Wyoming            Stub   
20154       140245     Worland, Wyoming  Wyoming            Stub   
20155       140070      Wright, Wyoming  Wyoming           Start   
20156       140112       Yoder, Wyoming  Wyoming           Start   

                                                     url  
0       https://en.wikipedia.org/wiki/Abbeville,_Alabama  
1      https://en.wikipedia.org/wiki/Adamsville,_Alabama  
2         https://en.wikipedia.org/wiki/Addison,_Alabama  
3           https://en.wikipedia.org/wiki/Akron,_Alabama  
4       https://en.wikipedia.org/wiki/Alabaster,_Alabama  
...                                                  ...  
20152   https://en.wikipedia.org/wiki/Wamsutter,_Wyoming  
20153   https://en.wikipedia.org/wiki/Wheatland,_Wyoming  
20154     https://en.wikipedia.org/wiki/Worland,_Wyoming  
20155      https://en.wikipedia.org/wiki/Wright,_Wyoming  
20156       https://en.wikipedia.org/wiki/Yoder,_Wyoming  

[20157 rows x 5 columns]

The NST population excel file is quite messy and not properly formatted. Hence, I specify an unusual row for the header and drop the first 5 rows of the data (because they are not states).

In [3]:
df_NST_pop = pd.read_excel('NST-EST2022-POP.xlsx', header=3)
df_NST_pop = df_NST_pop.iloc[5:, :]

# Source: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.reset_index.html
# Source: https://stackoverflow.com/questions/16167829/in-pandas-how-can-i-reset-index-without-adding-a-new-column
df_NST_pop.reset_index(drop=True, inplace=True)
df_NST_pop

Unnamed: 0  Unnamed: 1        2020  \
0                                            .Alabama   5024356.0   5031362.0   
1                                             .Alaska    733378.0    732923.0   
2                                            .Arizona   7151507.0   7179943.0   
3                                           .Arkansas   3011555.0   3014195.0   
4                                         .California  39538245.0  39501653.0   
5                                           .Colorado   5773733.0   5784865.0   
6                                        .Connecticut   3605942.0   3597362.0   
7                                           .Delaware    989957.0    992114.0   
8                               .District of Columbia    689546.0    670868.0   
9                                            .Florida  21538226.0  21589602.0   
10                                           .Georgia  10711937.0  10729828.0   
11                                            .Hawaii   1455273.0   1451043.0   
12                                             .Idaho   1839092.0   1849202.0   
13                                          .Illinois  12812545.0  12786580.0   
14                                           .Indiana   6785668.0   6788799.0   
15                                              .Iowa   3190372.0   3190571.0   
16                                            .Kansas   2937847.0   2937919.0   
17                                          .Kentucky   4505893.0   4507445.0   
18                                         .Louisiana   4657749.0   4651664.0   
19                                             .Maine   1362341.0   1363557.0   
20                                          .Maryland   6177213.0   6173205.0   
21                                     .Massachusetts   7029949.0   6995729.0   
22                                          .Michigan  10077325.0  10069577.0   
23                                         .Minnesota   5706504.0   5709852.0   
24                                       .Mississippi   2961288.0   2958141.0   
25                                          .Missouri   6154920.0   6153998.0   
26                                           .Montana   1084197.0   1087075.0   
27                                          .Nebraska   1961489.0   1962642.0   
28                                            .Nevada   3104624.0   3115648.0   
29                                     .New Hampshire   1377518.0   1378587.0   
30                                        .New Jersey   9289031.0   9271689.0   
31                                        .New Mexico   2117527.0   2118390.0   
32                                          .New York  20201230.0  20108296.0   
33                                    .North Carolina  10439414.0  10449445.0   
34                                      .North Dakota    779091.0    779518.0   
35                                              .Ohio  11799374.0  11797517.0   
36                                          .Oklahoma   3959346.0   3964912.0   
37                                            .Oregon   4237291.0   4244795.0   
38                                      .Pennsylvania  13002689.0  12994440.0   
39                                      .Rhode Island   1097371.0   1096345.0   
40                                    .South Carolina   5118429.0   5131848.0   
41                                      .South Dakota    886677.0    887799.0   
42                                         .Tennessee   6910786.0   6925619.0   
43                                             .Texas  29145428.0  29232474.0   
44                                              .Utah   3271614.0   3283785.0   
45                                           .Vermont    643085.0    642893.0   
46                                          .Virginia   8631384.0   8636471.0   
47                                        .Washington   7705247.0   7724031.0   
48                                     .West Virginia   1793755.0   179142

Now, from the assignment instructions, we can infer that we only need to concern ourselves with the state column and the 2022 population data, so let's drop some columns and do some labeling. Also, notice that for some reason all the state values begin with a '.', so let's get rid of those. Lastly, let's drop the last few rows of the table which are not states

In [4]:
# Source: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop.html
df_NST_pop.drop(['Unnamed: 1', 2020, 2021], axis=1, inplace=True)

# Source: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rename.html
# Source: https://www.geeksforgeeks.org/how-to-rename-columns-in-pandas-dataframe/
df_NST_pop.rename(columns={'Unnamed: 0': 'state', 2022 : 'population'}, inplace=True)

# I'm not really sure how this works, but I got the following approach from
# https://stackoverflow.com/questions/42349572/remove-first-x-number-of-characters-from-each-row-in-a-column-of-a-python-datafr
df_NST_pop['state'] = df_NST_pop['state'].str[1:]

# Source: https://www.geeksforgeeks.org/remove-last-n-rows-of-a-pandas-dataframe/
df_NST_pop.drop(df_NST_pop.tail(7).index, inplace=True)

df_NST_pop

state  population
0                Alabama   5074296.0
1                 Alaska    733583.0
2                Arizona   7359197.0
3               Arkansas   3045637.0
4             California  39029342.0
5               Colorado   5839926.0
6            Connecticut   3626205.0
7               Delaware   1018396.0
8   District of Columbia    671803.0
9                Florida  22244823.0
10               Georgia  10912876.0
11                Hawaii   1440196.0
12                 Idaho   1939033.0
13              Illinois  12582032.0
14               Indiana   6833037.0
15                  Iowa   3200517.0
16                Kansas   2937150.0
17              Kentucky   4512310.0
18             Louisiana   4590241.0
19                 Maine   1385340.0
20              Maryland   6164660.0
21         Massachusetts   6981974.0
22              Michigan  10034113.0
23             Minnesota   5717184.0
24           Mississippi   2940057.0
25              Missouri   6177957.0
26               Montana   1122867.0
27              Nebraska   1967923.0
28                Nevada   3177772.0
29         New Hampshire   1395231.0
30            New Jersey   9261699.0
31            New Mexico   2113344.0
32              New York  19677151.0
33        North Carolina  10698973.0
34          North Dakota    779261.0
35                  Ohio  11756058.0
36              Oklahoma   4019800.0
37                Oregon   4240137.0
38          Pennsylvania  12972008.0
39          Rhode Island   1093734.0
40        South Carolina   5282634.0
41          South Dakota    909824.0
42             Tennessee   7051339.0
43                 Texas  30029572.0
44                  Utah   3380800.0
45               Vermont    647064.0
46              Virginia   8683619.0
47            Washington   7785786.0
48         West Virginia   1775156.0
49             Wisconsin   5892539.0
50               Wyoming    581381.0

Now we can merge the two tables and rename the 'page_title' column to 'article_title' to match the schema in the assignment instructions

In [5]:
# Source: https://towardsdatascience.com/left-join-with-pandas-data-frames-in-python-c29c85089ba4
# Source: https://stackoverflow.com/questions/53645882/pandas-merging-101
df_quality_pop = df_quality.merge(df_NST_pop, on='state', how='inner')

# Rearrange the column order
df_quality_pop = df_quality_pop[['revision_id', 'page_title', 'state', 'population', 'article_quality', 'url']]

# Rename the 'page_title' column
df_quality_pop.rename(columns = {'page_title': 'article_title'}, inplace=True)

df_quality_pop

revision_id        article_title    state  population article_quality  \
0           104730   Abbeville, Alabama  Alabama   5074296.0            Stub   
1           104761  Adamsville, Alabama  Alabama   5074296.0            Stub   
2           105188     Addison, Alabama  Alabama   5074296.0            Stub   
3           104726       Akron, Alabama  Alabama   5074296.0            Stub   
4           105109   Alabaster, Alabama  Alabama   5074296.0            Stub   
...            ...                  ...      ...         ...             ...   
17030       140221   Wamsutter, Wyoming  Wyoming    581381.0           Start   
17031       140185   Wheatland, Wyoming  Wyoming    581381.0            Stub   
17032       140245     Worland, Wyoming  Wyoming    581381.0            Stub   
17033       140070      Wright, Wyoming  Wyoming    581381.0           Start   
17034       140112       Yoder, Wyoming  Wyoming    581381.0           Start   

                                                     url  
0       https://en.wikipedia.org/wiki/Abbeville,_Alabama  
1      https://en.wikipedia.org/wiki/Adamsville,_Alabama  
2         https://en.wikipedia.org/wiki/Addison,_Alabama  
3           https://en.wikipedia.org/wiki/Akron,_Alabama  
4       https://en.wikipedia.org/wiki/Alabaster,_Alabama  
...                                                  ...  
17030   https://en.wikipedia.org/wiki/Wamsutter,_Wyoming  
17031   https://en.wikipedia.org/wiki/Wheatland,_Wyoming  
17032     https://en.wikipedia.org/wiki/Worland,_Wyoming  
17033      https://en.wikipedia.org/wiki/Wright,_Wyoming  
17034       https://en.wikipedia.org/wiki/Yoder,_Wyoming  

[17035 rows x 6 columns]

**Note:** Importantly, before the merge we had 20157 rows of data in our article quality table. However, after the merge we only have 17035. Why is that? 

Let's identify all states in the table of data I scraped that are not found in the population table

In [6]:
# Source: https://www.geeksforgeeks.org/check-if-element-exists-in-list-in-python/#

for state in df_quality.state.unique():
    if state not in df_NST_pop.state.unique():
        print("{} not found in NST population table".format(state))

Georgia_(U.S._state) not found in NST population table
New_Hampshire not found in NST population table
New_Jersey not found in NST population table
New_Mexico not found in NST population table
New_York not found in NST population table
North_Carolina not found in NST population table
North_Dakota not found in NST population table
Rhode_Island not found in NST population table
South_Carolina not found in NST population table
South_Dakota not found in NST population table
West_Virginia not found in NST population table


By inspection we note that the majority of the discrepancies are caused by a difference in how two-word state names are handled. For consistancy, let's replace the spaces with underscores in the population table. Additionally, let's change the name for Georgia in the article quality table from 'Georgia_(U.S._state)' to 'Georgia'.

In [7]:
# Source: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.replace.html
df_quality.state.replace('Georgia_(U.S._state)', 'Georgia', inplace=True)

# Source: https://stackoverflow.com/questions/42462530/how-to-replace-the-white-space-in-a-string-in-a-pandas-dataframe. 
# I am not sure why the regex argument needs to be true, though
df_NST_pop.state.replace(' ', '_', regex=True, inplace=True)

Now, let's confirm that all states in the article quality table are found in the NST population table.

In [8]:
# Source: https://www.geeksforgeeks.org/check-if-element-exists-in-list-in-python/#

for state in df_quality.state.unique():
    if state not in df_NST_pop.state.unique():
        print("{} not found in NST population table".format(state))

Since there are no printed results, we can assume the column values are ready to be merged. Once again we try merging the two tables and get the following result, with the full 20,157 rows.

In [9]:
# Source: https://towardsdatascience.com/left-join-with-pandas-data-frames-in-python-c29c85089ba4
# Source: https://stackoverflow.com/questions/53645882/pandas-merging-101
df_quality_pop = df_quality.merge(df_NST_pop, on='state', how='inner')

# Rearrange the column order
df_quality_pop = df_quality_pop[['revision_id', 'page_title', 'state', 'population', 'article_quality', 'url']]

# Rename the 'page_title' column
df_quality_pop.rename(columns = {'page_title': 'article_title'}, inplace=True)

df_quality_pop

revision_id        article_title    state  population article_quality  \
0           104730   Abbeville, Alabama  Alabama   5074296.0            Stub   
1           104761  Adamsville, Alabama  Alabama   5074296.0            Stub   
2           105188     Addison, Alabama  Alabama   5074296.0            Stub   
3           104726       Akron, Alabama  Alabama   5074296.0            Stub   
4           105109   Alabaster, Alabama  Alabama   5074296.0            Stub   
...            ...                  ...      ...         ...             ...   
20152       140221   Wamsutter, Wyoming  Wyoming    581381.0           Start   
20153       140185   Wheatland, Wyoming  Wyoming    581381.0            Stub   
20154       140245     Worland, Wyoming  Wyoming    581381.0            Stub   
20155       140070      Wright, Wyoming  Wyoming    581381.0           Start   
20156       140112       Yoder, Wyoming  Wyoming    581381.0           Start   

                                                     url  
0       https://en.wikipedia.org/wiki/Abbeville,_Alabama  
1      https://en.wikipedia.org/wiki/Adamsville,_Alabama  
2         https://en.wikipedia.org/wiki/Addison,_Alabama  
3           https://en.wikipedia.org/wiki/Akron,_Alabama  
4       https://en.wikipedia.org/wiki/Alabaster,_Alabama  
...                                                  ...  
20152   https://en.wikipedia.org/wiki/Wamsutter,_Wyoming  
20153   https://en.wikipedia.org/wiki/Wheatland,_Wyoming  
20154     https://en.wikipedia.org/wiki/Worland,_Wyoming  
20155      https://en.wikipedia.org/wiki/Wright,_Wyoming  
20156       https://en.wikipedia.org/wiki/Yoder,_Wyoming  

[20157 rows x 6 columns]

**Now, let's merge in the geographic region corresponding to each state:**

First, let's import the geographical data. According to a classmate, she was told to focus on the 'DIVISION' column - so we can ignore the 'REGION' column. Let's also rename the columns

In [10]:
df_region = pd.read_excel('US States by Region - US Census Bureau.xlsx')

# Source: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop.html
df_region.drop(['REGION'], axis=1, inplace=True)

# Source: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rename.html
# Source: https://www.geeksforgeeks.org/how-to-rename-columns-in-pandas-dataframe/
df_region.rename(columns={'DIVISION': 'regional_division', 'STATE' : 'state'}, inplace=True)

df_region.head(10)

regional_division          state
0               NaN            NaN
1       New England            NaN
2               NaN    Connecticut
3               NaN          Maine
4               NaN  Massachusetts
5               NaN  New Hampshire
6               NaN   Rhode Island
7               NaN        Vermont
8   Middle Atlantic            NaN
9               NaN     New Jersey

Now, we need to fill in the DIVISION column so that it has no missing values, to make the merge easier

In [11]:
# Source: https://stackoverflow.com/questions/16476924/how-to-iterate-over-rows-in-a-dataframe-in-pandas
# Source: used this site to learn that .notna() is not a function for floats and what to do instead
# https://stackoverflow.com/questions/13413590/how-to-drop-rows-of-pandas-dataframe-whose-value-in-a-certain-column-is-nan
# Source: https://pandas.pydata.org/docs/reference/api/pandas.notna.html
current_division = np.nan

for index, row in df_region.iterrows():
    if pd.notna(row.regional_division):
        current_division = row.regional_division
    
    else:
        row.regional_division = current_division

df_region.head(10)

regional_division          state
0               NaN            NaN
1       New England            NaN
2       New England    Connecticut
3       New England          Maine
4       New England  Massachusetts
5       New England  New Hampshire
6       New England   Rhode Island
7       New England        Vermont
8   Middle Atlantic            NaN
9   Middle Atlantic     New Jersey

Next, we can drop the null values

In [12]:
df_region.dropna(inplace=True)
df_region.head()

regional_division          state
2       New England    Connecticut
3       New England          Maine
4       New England  Massachusetts
5       New England  New Hampshire
6       New England   Rhode Island

Finally, we can merge this dataset with the other dataframe we have and rearrange the columns to match the order they appear in the assignment schema

In [13]:
# Source: https://towardsdatascience.com/left-join-with-pandas-data-frames-in-python-c29c85089ba4
# Source: https://stackoverflow.com/questions/53645882/pandas-merging-101
df_quality_pop_division = df_quality_pop.merge(df_region, on='state', how='inner')

df_quality_pop_division = df_quality_pop_division[['state', 'regional_division', 'population', 'article_title', 
                                                   'revision_id', 'article_quality']]
df_quality_pop_division

state   regional_division  population        article_title  \
0      Alabama  East South Central   5074296.0   Abbeville, Alabama   
1      Alabama  East South Central   5074296.0  Adamsville, Alabama   
2      Alabama  East South Central   5074296.0     Addison, Alabama   
3      Alabama  East South Central   5074296.0       Akron, Alabama   
4      Alabama  East South Central   5074296.0   Alabaster, Alabama   
...        ...                 ...         ...                  ...   
17495  Wyoming            Mountain    581381.0   Wamsutter, Wyoming   
17496  Wyoming            Mountain    581381.0   Wheatland, Wyoming   
17497  Wyoming            Mountain    581381.0     Worland, Wyoming   
17498  Wyoming            Mountain    581381.0      Wright, Wyoming   
17499  Wyoming            Mountain    581381.0       Yoder, Wyoming   

       revision_id article_quality  
0           104730            Stub  
1           104761            Stub  
2           105188            Stub  
3           104726            Stub  
4           105109            Stub  
...            ...             ...  
17495       140221           Start  
17496       140185            Stub  
17497       140245            Stub  
17498       140070           Start  
17499       140112           Start  

[17500 rows x 6 columns]

Once again we see that instead of the 20157 rows, our merged data contains only 17500, suggesting there were issues with the merge. Let's once again check the discrepancies.

In [14]:
# Source: https://www.geeksforgeeks.org/check-if-element-exists-in-list-in-python/#

for state in df_quality_pop.state.unique():
    if state not in df_region.state.unique():
        print("{} not found in NST population table".format(state))

New_Hampshire not found in NST population table
New_Jersey not found in NST population table
New_Mexico not found in NST population table
New_York not found in NST population table
North_Carolina not found in NST population table
North_Dakota not found in NST population table
Rhode_Island not found in NST population table
South_Carolina not found in NST population table
South_Dakota not found in NST population table
West_Virginia not found in NST population table


Once again we see that there is probably an inconsistancy in how the two-name states are represented. Let's fix this by renaming the two-word state names in the region table, and then verify that this resolved the issue.

In [15]:
# Source: https://stackoverflow.com/questions/42462530/how-to-replace-the-white-space-in-a-string-in-a-pandas-dataframe. 
# I am not sure why the regex argument needs to be true, though
df_region.state.replace(' ', '_', regex=True, inplace=True)

# Source: https://towardsdatascience.com/left-join-with-pandas-data-frames-in-python-c29c85089ba4
# Source: https://stackoverflow.com/questions/53645882/pandas-merging-101
df_quality_pop_division = df_quality_pop.merge(df_region, on='state', how='inner')

df_quality_pop_division = df_quality_pop_division[['state', 'regional_division', 'population', 'article_title', 
                                                   'revision_id', 'article_quality']]
df_quality_pop_division

state   regional_division  population        article_title  \
0      Alabama  East South Central   5074296.0   Abbeville, Alabama   
1      Alabama  East South Central   5074296.0  Adamsville, Alabama   
2      Alabama  East South Central   5074296.0     Addison, Alabama   
3      Alabama  East South Central   5074296.0       Akron, Alabama   
4      Alabama  East South Central   5074296.0   Alabaster, Alabama   
...        ...                 ...         ...                  ...   
20152  Wyoming            Mountain    581381.0   Wamsutter, Wyoming   
20153  Wyoming            Mountain    581381.0   Wheatland, Wyoming   
20154  Wyoming            Mountain    581381.0     Worland, Wyoming   
20155  Wyoming            Mountain    581381.0      Wright, Wyoming   
20156  Wyoming            Mountain    581381.0       Yoder, Wyoming   

       revision_id article_quality  
0           104730            Stub  
1           104761            Stub  
2           105188            Stub  
3           104726            Stub  
4           105109            Stub  
...            ...             ...  
20152       140221           Start  
20153       140185            Stub  
20154       140245            Stub  
20155       140070           Start  
20156       140112           Start  

[20157 rows x 6 columns]

We now have the expected set of results, implying our solution resolved the merge issue. Now, let's save the data to a csv as requested.

In [16]:
df_quality_pop_division.to_csv('wp_scored_city_articles_by_state.csv', index=False)

# Step 4: Analysis

For the analysis, we will compute the total articles per capita for the states and regional divisions in our data and then in step 5 we will rank the results.

In [17]:
# Source: https://www.geeksforgeeks.org/python-pandas-dataframe-groupby/
# Source: https://stackoverflow.com/questions/10373660/converting-a-pandas-groupby-output-from-series-to-dataframe
# https://stackoverflow.com/questions/32751229/pandas-sum-by-groupby-but-exclude-certain-columns

df_state_article_counts = df_quality_pop_division.groupby(['state', 'population'], as_index=False)[['article_title']].agg('count')
df_state_article_counts

state  population  article_title
0          Alabama   5074296.0            443
1           Alaska    733583.0            142
2          Arizona   7359197.0             89
3         Arkansas   3045637.0            482
4       California  39029342.0            458
5         Colorado   5839926.0            258
6         Delaware   1018396.0             50
7          Florida  22244823.0            393
8          Georgia  10912876.0            465
9           Hawaii   1440196.0            130
10           Idaho   1939033.0            181
11        Illinois  12582032.0           1226
12         Indiana   6833037.0            551
13            Iowa   3200517.0            999
14          Kansas   2937150.0             59
15        Kentucky   4512310.0            393
16       Louisiana   4590241.0            270
17           Maine   1385340.0            430
18        Maryland   6164660.0            152
19   Massachusetts   6981974.0            326
20        Michigan  10034113.0           1653
21       Minnesota   5717184.0            803
22     Mississippi   2940057.0            281
23        Missouri   6177957.0            888
24         Montana   1122867.0            118
25          Nevada   3177772.0             18
26   New_Hampshire   1395231.0            219
27      New_Jersey   9261699.0            533
28      New_Mexico   2113344.0             94
29        New_York  19677151.0            616
30  North_Carolina  10698973.0             47
31    North_Dakota    779261.0            337
32            Ohio  11756058.0            882
33        Oklahoma   4019800.0             73
34          Oregon   4240137.0            233
35    Pennsylvania  12972008.0           2389
36    Rhode_Island   1093734.0             37
37  South_Carolina   5282634.0            253
38    South_Dakota    909824.0            296
39       Tennessee   7051339.0            327
40           Texas  30029572.0           1136
41            Utah   3380800.0            242
42         Vermont    647064.0            302
43        Virginia   8683619.0            129
44      Washington   7785786.0            254
45   West_Virginia   1775156.0            225
46       Wisconsin   5892539.0            179
47         Wyoming    581381.0             96

Now, let's do the math

In [18]:
df_articles_per_capita_state = df_state_article_counts
df_articles_per_capita_state['articles_per_capita_state'] = df_state_article_counts.article_title / df_state_article_counts.population
df_articles_per_capita_state

state  population  article_title  articles_per_capita_state
0          Alabama   5074296.0            443                   0.000087
1           Alaska    733583.0            142                   0.000194
2          Arizona   7359197.0             89                   0.000012
3         Arkansas   3045637.0            482                   0.000158
4       California  39029342.0            458                   0.000012
5         Colorado   5839926.0            258                   0.000044
6         Delaware   1018396.0             50                   0.000049
7          Florida  22244823.0            393                   0.000018
8          Georgia  10912876.0            465                   0.000043
9           Hawaii   1440196.0            130                   0.000090
10           Idaho   1939033.0            181                   0.000093
11        Illinois  12582032.0           1226                   0.000097
12         Indiana   6833037.0            551                   0.000081
13            Iowa   3200517.0            999                   0.000312
14          Kansas   2937150.0             59                   0.000020
15        Kentucky   4512310.0            393                   0.000087
16       Louisiana   4590241.0            270                   0.000059
17           Maine   1385340.0            430                   0.000310
18        Maryland   6164660.0            152                   0.000025
19   Massachusetts   6981974.0            326                   0.000047
20        Michigan  10034113.0           1653                   0.000165
21       Minnesota   5717184.0            803                   0.000140
22     Mississippi   2940057.0            281                   0.000096
23        Missouri   6177957.0            888                   0.000144
24         Montana   1122867.0            118                   0.000105
25          Nevada   3177772.0             18                   0.000006
26   New_Hampshire   1395231.0            219                   0.000157
27      New_Jersey   9261699.0            533                   0.000058
28      New_Mexico   2113344.0             94                   0.000044
29        New_York  19677151.0            616                   0.000031
30  North_Carolina  10698973.0             47                   0.000004
31    North_Dakota    779261.0            337                   0.000432
32            Ohio  11756058.0            882                   0.000075
33        Oklahoma   4019800.0             73                   0.000018
34          Oregon   4240137.0            233                   0.000055
35    Pennsylvania  12972008.0           2389                   0.000184
36    Rhode_Island   1093734.0             37                   0.000034
37  South_Carolina   5282634.0            253                   0.000048
38    South_Dakota    909824.0            296                   0.000325
39       Tennessee   7051339.0            327                   0.000046
40           Texas  30029572.0           1136                   0.000038
41            Utah   3380800.0            242                   0.000072
42         Vermont    647064.0            302                   0.000467
43        Virginia   8683619.0            129                   0.000015
44      Washington   7785786.0            254                   0.000033
45   West_Virginia   1775156.0            225                   0.000127
46       Wisconsin   5892539.0            179                   0.000030
47         Wyoming    581381.0             96                   0.000165

**Now we compute the total articles per capita for the regional divisions**

First we count the number of articles in each state

In [19]:
# Source: https://www.geeksforgeeks.org/python-pandas-dataframe-groupby/
# Source: https://stackoverflow.com/questions/10373660/converting-a-pandas-groupby-output-from-series-to-dataframe
# https://stackoverflow.com/questions/32751229/pandas-sum-by-groupby-but-exclude-certain-columns

df_division_article_counts = df_quality_pop_division.groupby(['regional_division', 'state', 'population'], as_index=False)[['article_title']].agg('count')
df_division_article_counts

regional_division           state  population  article_title
0   East North Central        Illinois  12582032.0           1226
1   East North Central         Indiana   6833037.0            551
2   East North Central        Michigan  10034113.0           1653
3   East North Central            Ohio  11756058.0            882
4   East North Central       Wisconsin   5892539.0            179
5   East South Central         Alabama   5074296.0            443
6   East South Central        Kentucky   4512310.0            393
7   East South Central     Mississippi   2940057.0            281
8   East South Central       Tennessee   7051339.0            327
9      Middle Atlantic      New_Jersey   9261699.0            533
10     Middle Atlantic        New_York  19677151.0            616
11     Middle Atlantic    Pennsylvania  12972008.0           2389
12            Mountain         Arizona   7359197.0             89
13            Mountain        Colorado   5839926.0            258
14            Mountain           Idaho   1939033.0            181
15            Mountain         Montana   1122867.0            118
16            Mountain          Nevada   3177772.0             18
17            Mountain      New_Mexico   2113344.0             94
18            Mountain            Utah   3380800.0            242
19            Mountain         Wyoming    581381.0             96
20         New England           Maine   1385340.0            430
21         New England   Massachusetts   6981974.0            326
22         New England   New_Hampshire   1395231.0            219
23         New England    Rhode_Island   1093734.0             37
24         New England         Vermont    647064.0            302
25             Pacific          Alaska    733583.0            142
26             Pacific      California  39029342.0            458
27             Pacific          Hawaii   1440196.0            130
28             Pacific          Oregon   4240137.0            233
29             Pacific      Washington   7785786.0            254
30      South Atlantic        Delaware   1018396.0             50
31      South Atlantic         Florida  22244823.0            393
32      South Atlantic         Georgia  10912876.0            465
33      South Atlantic        Maryland   6164660.0            152
34      South Atlantic  North_Carolina  10698973.0             47
35      South Atlantic  South_Carolina   5282634.0            253
36      South Atlantic        Virginia   8683619.0            129
37      South Atlantic   West_Virginia   1775156.0            225
38  West North Central            Iowa   3200517.0            999
39  West North Central          Kansas   2937150.0             59
40  West North Central       Minnesota   5717184.0            803
41  West North Central        Missouri   6177957.0            888
42  West North Central    North_Dakota    779261.0            337
43  West North Central    South_Dakota    909824.0            296
44  West South Central        Arkansas   3045637.0            482
45  West South Central       Louisiana   4590241.0            270
46  West South Central        Oklahoma   4019800.0             73
47  West South Central           Texas  30029572.0           1136

Then we add the population and article counts in each state to find the total population and article counts for each regional division.

In [20]:
# Source: https://www.geeksforgeeks.org/python-pandas-dataframe-groupby/
# Source: https://stackoverflow.com/questions/10373660/converting-a-pandas-groupby-output-from-series-to-dataframe
# https://stackoverflow.com/questions/32751229/pandas-sum-by-groupby-but-exclude-certain-columns

df_division_article_counts = df_division_article_counts.groupby(['regional_division'], as_index=False)[['population', 'article_title']].agg('sum')
df_division_article_counts

regional_division  population  article_title
0  East North Central  47097779.0           4491
1  East South Central  19578002.0           1444
2     Middle Atlantic  41910858.0           3538
3            Mountain  25514320.0           1096
4         New England  11503343.0           1314
5             Pacific  53229044.0           1217
6      South Atlantic  66781137.0           1714
7  West North Central  19721893.0           3382
8  West South Central  41685250.0           1961

Now we can compute the per capita values

In [21]:
df_articles_per_capita_region = df_division_article_counts
df_articles_per_capita_region['articles_per_capita_region'] = df_division_article_counts.article_title / df_division_article_counts.population

In [22]:
df_articles_per_capita_region

regional_division  population  article_title  articles_per_capita_region
0  East North Central  47097779.0           4491                    0.000095
1  East South Central  19578002.0           1444                    0.000074
2     Middle Atlantic  41910858.0           3538                    0.000084
3            Mountain  25514320.0           1096                    0.000043
4         New England  11503343.0           1314                    0.000114
5             Pacific  53229044.0           1217                    0.000023
6      South Atlantic  66781137.0           1714                    0.000026
7  West North Central  19721893.0           3382                    0.000171
8  West South Central  41685250.0           1961                    0.000047

**Article Quality:**

To determine the number of high-quality articles per state and region population (per capita) we start by determining whether or not each article is high-quality

In [23]:
df_high_quality = df_quality_pop_division

# I used the following sources to learn that np.where() existed and how to use it
# https://stackoverflow.com/questions/21702342/creating-a-new-column-based-on-if-elif-else-condition
# https://stackoverflow.com/questions/19913659/how-do-i-create-a-new-column-where-the-values-are-selected-based-on-existing-col?noredirect=1&lq=1
# https://numpy.org/doc/stable/reference/generated/numpy.where.html
# https://stackoverflow.com/questions/36921951/truth-value-of-a-series-is-ambiguous-use-a-empty-a-bool-a-item-a-any-o
# https://stackoverflow.com/questions/16343752/numpy-where-function-multiple-conditions


df_high_quality['is_high_quality'] = np.where((df_high_quality['article_quality'] == 'FA') 
                                                      | (df_high_quality['article_quality'] == 'GA'), 1, 0)
df_high_quality[df_high_quality['is_high_quality'] == 1]

state   regional_division  population  \
38           Alabama  East South Central   5074296.0   
1346      California             Pacific  39029342.0   
2756         Georgia      South Atlantic  10912876.0   
2926           Idaho            Mountain   1939033.0   
3980        Illinois  East North Central  12582032.0   
4102        Illinois  East North Central  12582032.0   
4700         Indiana  East North Central   6833037.0   
6901           Maine         New England   1385340.0   
6967           Maine         New England   1385340.0   
7052        Maryland      South Atlantic   6164660.0   
7491   Massachusetts         New England   6981974.0   
7875        Michigan  East North Central  10034113.0   
7892        Michigan  East North Central  10034113.0   
8624        Michigan  East North Central  10034113.0   
8630        Michigan  East North Central  10034113.0   
9091        Michigan  East North Central  10034113.0   
9100        Michigan  East North Central  10034113.0   
12141       New_York     Middle Atlantic  19677151.0   
12332       New_York     Middle Atlantic  19677151.0   
13232           Ohio  East North Central  11756058.0   
14350   Pennsylvania     Middle Atlantic  12972008.0   
14360   Pennsylvania     Middle Atlantic  12972008.0   
14597   Pennsylvania     Middle Atlantic  12972008.0   
14775   Pennsylvania     Middle Atlantic  12972008.0   
15026   Pennsylvania     Middle Atlantic  12972008.0   
15027   Pennsylvania     Middle Atlantic  12972008.0   
15267   Pennsylvania     Middle Atlantic  12972008.0   
15480   Pennsylvania     Middle Atlantic  12972008.0   
15759   Pennsylvania     Middle Atlantic  12972008.0   
18507          Texas  West South Central  30029572.0   
19200        Vermont         New England    647064.0   
19278       Virginia      South Atlantic   8683619.0   
19285       Virginia      South Atlantic   8683619.0   
19286       Virginia      South Atlantic   8683619.0   
19290       Virginia      South Atlantic   8683619.0   
19873  West_Virginia      South Atlantic   1775156.0   

                                           article_title  revision_id  \
38                                       Berlin, Alabama     23831085   
1346                                  Irvine, California      5201333   
2756                              Warner Robins, Georgia     35849769   
2926                                       Hayden, Idaho     31378210   
3980                                Plattville, Illinois     18407774   
4102                                  Shabbona, Illinois     25987812   
4700                                Millersburg, Indiana     15868841   
6901                                     Rangeley, Maine     35981352   
6967                                        Unity, Maine     46954769   
7052                               Church Hill, Maryland     43434329   
7491                             Winthrop, Massachusetts     28742759   
7875                   Cooper Charter Township, Michigan     22153539   
7892                    Crystal Falls Township, Michigan     28790113   
8624                       Northville Township, Michigan     58269693   
8630          Norwich Township, Newaygo County, Michigan     28790207   
9091                               West Branch, Michigan     28790086   
9100        Wheatland Township, Mecosta County, Michigan     28742718   
12141                                  Batavia, New York     47739945   
12332                                   Hudson, New York     26149742   
13232             Monroe, Butler and Warren County, Ohio     62801517   
14350      Moon Township, Allegheny County, Pennsylvania     74939907   
14360  Unity Township, Westmoreland County, Pennsylvania     74822494   
14597  Southampton Township, Franklin County, Pennsyl...     74939894   
14775                         Shippensburg, Pennsylvania     74939825   
15026      Jackson Township, Butler County, Pennsylvania     74822297   
15027                           Parkesburg, 

We can see the high-quality results in the output directly above. Now, we can group by the states and regions to find the per capita high-quality articles

In [24]:
df_high_quality_count = df_high_quality.groupby(['regional_division', 'state', 'population'], as_index=False)['is_high_quality'].agg('sum')
df_high_quality_count

regional_division           state  population  is_high_quality
0   East North Central        Illinois  12582032.0                2
1   East North Central         Indiana   6833037.0                1
2   East North Central        Michigan  10034113.0                6
3   East North Central            Ohio  11756058.0                1
4   East North Central       Wisconsin   5892539.0                0
5   East South Central         Alabama   5074296.0                1
6   East South Central        Kentucky   4512310.0                0
7   East South Central     Mississippi   2940057.0                0
8   East South Central       Tennessee   7051339.0                0
9      Middle Atlantic      New_Jersey   9261699.0                0
10     Middle Atlantic        New_York  19677151.0                2
11     Middle Atlantic    Pennsylvania  12972008.0                9
12            Mountain         Arizona   7359197.0                0
13            Mountain        Colorado   5839926.0                0
14            Mountain           Idaho   1939033.0                1
15            Mountain         Montana   1122867.0                0
16            Mountain          Nevada   3177772.0                0
17            Mountain      New_Mexico   2113344.0                0
18            Mountain            Utah   3380800.0                0
19            Mountain         Wyoming    581381.0                0
20         New England           Maine   1385340.0                2
21         New England   Massachusetts   6981974.0                1
22         New England   New_Hampshire   1395231.0                0
23         New England    Rhode_Island   1093734.0                0
24         New England         Vermont    647064.0                1
25             Pacific          Alaska    733583.0                0
26             Pacific      California  39029342.0                1
27             Pacific          Hawaii   1440196.0                0
28             Pacific          Oregon   4240137.0                0
29             Pacific      Washington   7785786.0                0
30      South Atlantic        Delaware   1018396.0                0
31      South Atlantic         Florida  22244823.0                0
32      South Atlantic         Georgia  10912876.0                1
33      South Atlantic        Maryland   6164660.0                1
34      South Atlantic  North_Carolina  10698973.0                0
35      South Atlantic  South_Carolina   5282634.0                0
36      South Atlantic        Virginia   8683619.0                4
37      South Atlantic   West_Virginia   1775156.0                1
38  West North Central            Iowa   3200517.0                0
39  West North Central          Kansas   2937150.0                0
40  West North Central       Minnesota   5717184.0                0
41  West North Central        Missouri   6177957.0                0
42  West North Central    North_Dakota    779261.0                0
43  West North Central    South_Dakota    909824.0                0
44  West South Central        Arkansas   3045637.0                0
45  West South Central       Louisiana   4590241.0                0
46  West South Central        Oklahoma   4019800.0                0
47  West South Central           Texas  30029572.0                1

Next, we find the high-quality articles per capita for states

In [25]:
df_high_quality_state = df_high_quality_count.drop('regional_division', axis=1)
df_high_quality_state['high_quality_per_capita'] = df_high_quality_state['is_high_quality'] / df_high_quality_state['population']
df_high_quality_state

state  population  is_high_quality  high_quality_per_capita
0         Illinois  12582032.0                2             1.589568e-07
1          Indiana   6833037.0                1             1.463478e-07
2         Michigan  10034113.0                6             5.979602e-07
3             Ohio  11756058.0                1             8.506253e-08
4        Wisconsin   5892539.0                0             0.000000e+00
5          Alabama   5074296.0                1             1.970717e-07
6         Kentucky   4512310.0                0             0.000000e+00
7      Mississippi   2940057.0                0             0.000000e+00
8        Tennessee   7051339.0                0             0.000000e+00
9       New_Jersey   9261699.0                0             0.000000e+00
10        New_York  19677151.0                2             1.016407e-07
11    Pennsylvania  12972008.0                9             6.938016e-07
12         Arizona   7359197.0                0             0.000000e+00
13        Colorado   5839926.0                0             0.000000e+00
14           Idaho   1939033.0                1             5.157210e-07
15         Montana   1122867.0                0             0.000000e+00
16          Nevada   3177772.0                0             0.000000e+00
17      New_Mexico   2113344.0                0             0.000000e+00
18            Utah   3380800.0                0             0.000000e+00
19         Wyoming    581381.0                0             0.000000e+00
20           Maine   1385340.0                2             1.443689e-06
21   Massachusetts   6981974.0                1             1.432260e-07
22   New_Hampshire   1395231.0                0             0.000000e+00
23    Rhode_Island   1093734.0                0             0.000000e+00
24         Vermont    647064.0                1             1.545442e-06
25          Alaska    733583.0                0             0.000000e+00
26      California  39029342.0                1             2.562175e-08
27          Hawaii   1440196.0                0             0.000000e+00
28          Oregon   4240137.0                0             0.000000e+00
29      Washington   7785786.0                0             0.000000e+00
30        Delaware   1018396.0                0             0.000000e+00
31         Florida  22244823.0                0             0.000000e+00
32         Georgia  10912876.0                1             9.163487e-08
33        Maryland   6164660.0                1             1.622149e-07
34  North_Carolina  10698973.0                0             0.000000e+00
35  South_Carolina   5282634.0                0             0.000000e+00
36        Virginia   8683619.0                4             4.606374e-07
37   West_Virginia   1775156.0                1             5.633308e-07
38            Iowa   3200517.0                0             0.000000e+00
39          Kansas   2937150.0                0             0.000000e+00
40       Minnesota   5717184.0                0             0.000000e+00
41        Missouri   6177957.0                0             0.000000e+00
42    North_Dakota    779261.0                0             0.000000e+00
43    South_Dakota    909824.0                0             0.000000e+00
44        Arkansas   3045637.0                0             0.000000e+00
45       Louisiana   4590241.0                0             0.000000e+00
46        Oklahoma   4019800.0                0             0.000000e+00
47           Texas  30029572.0                1             3.330051e-08

Lastly, we find the high-quality articles per capita for regions. We start by grouping the data by region

In [26]:
df_high_quality_region = df_high_quality_count.groupby('regional_division', as_index=False)['population', 'is_high_quality'].agg('sum')
df_high_quality_region

/tmp/ipykernel_9897/3704886650.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  df_high_quality_region = df_high_quality_count.groupby('regional_division', as_index=False)['population', 'is_high_quality'].agg('sum')


regional_division  population  is_high_quality
0  East North Central  47097779.0               10
1  East South Central  19578002.0                1
2     Middle Atlantic  41910858.0               11
3            Mountain  25514320.0                1
4         New England  11503343.0                4
5             Pacific  53229044.0                1
6      South Atlantic  66781137.0                7
7  West North Central  19721893.0                0
8  West South Central  41685250.0                1

Now we compute the per capita values

In [27]:
df_high_quality_region['high_quality_per_capita'] = df_high_quality_region['is_high_quality'] / df_high_quality_region['population']
df_high_quality_region

regional_division  population  is_high_quality  high_quality_per_capita
0  East North Central  47097779.0               10             2.123242e-07
1  East South Central  19578002.0                1             5.107774e-08
2     Middle Atlantic  41910858.0               11             2.624618e-07
3            Mountain  25514320.0                1             3.919368e-08
4         New England  11503343.0                4             3.477250e-07
5             Pacific  53229044.0                1             1.878674e-08
6      South Atlantic  66781137.0                7             1.048200e-07
7  West North Central  19721893.0                0             0.000000e+00
8  West South Central  41685250.0                1             2.398930e-08

# Step 5: Results

According to the assignment instructions, we are asked to produce 6 tables depicting the results of this analysis. They are listed in bold below and are direct quotes from the assignment instructions.

**1.	Top 10 US states by coverage: The 10 US states with the highest total articles per capita (in descending order)**.

In [28]:
# Source: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html
table_1 = df_articles_per_capita_state.sort_values(by='articles_per_capita_state', axis=0, ascending=False, ignore_index=True)
table_1.head(10)

state  population  article_title  articles_per_capita_state
0       Vermont    647064.0            302                   0.000467
1  North_Dakota    779261.0            337                   0.000432
2  South_Dakota    909824.0            296                   0.000325
3          Iowa   3200517.0            999                   0.000312
4         Maine   1385340.0            430                   0.000310
5        Alaska    733583.0            142                   0.000194
6  Pennsylvania  12972008.0           2389                   0.000184
7       Wyoming    581381.0             96                   0.000165
8      Michigan  10034113.0           1653                   0.000165
9      Arkansas   3045637.0            482                   0.000158

**2.	Bottom 10 US states by coverage: The 10 US states with the lowest total articles per capita (in ascending order)**.

In [29]:
# Source: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html
table_2 = df_articles_per_capita_state.sort_values(by='articles_per_capita_state', axis=0, ascending=True, ignore_index=True)
table_2.head(10)

state  population  article_title  articles_per_capita_state
0  North_Carolina  10698973.0             47                   0.000004
1          Nevada   3177772.0             18                   0.000006
2      California  39029342.0            458                   0.000012
3         Arizona   7359197.0             89                   0.000012
4        Virginia   8683619.0            129                   0.000015
5         Florida  22244823.0            393                   0.000018
6        Oklahoma   4019800.0             73                   0.000018
7          Kansas   2937150.0             59                   0.000020
8        Maryland   6164660.0            152                   0.000025
9       Wisconsin   5892539.0            179                   0.000030

**3.	Top 10 US states by high quality: The 10 US states with the highest high quality articles per capita (in descending order)**.

In [30]:
# Source: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html
table_3 = df_high_quality_state.sort_values(by='high_quality_per_capita', axis=0, ascending=False, ignore_index=True)
table_3.head(10)

state  population  is_high_quality  high_quality_per_capita
0        Vermont    647064.0                1             1.545442e-06
1          Maine   1385340.0                2             1.443689e-06
2   Pennsylvania  12972008.0                9             6.938016e-07
3       Michigan  10034113.0                6             5.979602e-07
4  West_Virginia   1775156.0                1             5.633308e-07
5          Idaho   1939033.0                1             5.157210e-07
6       Virginia   8683619.0                4             4.606374e-07
7        Alabama   5074296.0                1             1.970717e-07
8       Maryland   6164660.0                1             1.622149e-07
9       Illinois  12582032.0                2             1.589568e-07

**4.	Bottom 10 US states by high quality: The 10 US states with the lowest high quality articles per capita (in ascending order).**

Note: there are actually 31 states that have no high-quality articles. Hence, while I am only displaying 10 below, for the sake of the instructions, know that the 10 depicted are no worse than the other 21 states with no high-quality articles.

In [31]:
# Source: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html
table_4 = df_high_quality_state.sort_values(by='high_quality_per_capita', axis=0, ascending=True, ignore_index=True)
table_4.head(10)

state  population  is_high_quality  high_quality_per_capita
0    Rhode_Island   1093734.0                0                      0.0
1      Washington   7785786.0                0                      0.0
2          Oregon   4240137.0                0                      0.0
3          Hawaii   1440196.0                0                      0.0
4          Alaska    733583.0                0                      0.0
5  North_Carolina  10698973.0                0                      0.0
6        Oklahoma   4019800.0                0                      0.0
7   New_Hampshire   1395231.0                0                      0.0
8  South_Carolina   5282634.0                0                      0.0
9            Iowa   3200517.0                0                      0.0

**5.	Census divisions by total coverage: A rank ordered list of US census divisions (in descending order) by total articles per capita.**

In [32]:
# Source: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html
table_5 = df_articles_per_capita_region.sort_values(by='articles_per_capita_region', axis=0, ascending=False, ignore_index=True)
table_5

regional_division  population  article_title  articles_per_capita_region
0  West North Central  19721893.0           3382                    0.000171
1         New England  11503343.0           1314                    0.000114
2  East North Central  47097779.0           4491                    0.000095
3     Middle Atlantic  41910858.0           3538                    0.000084
4  East South Central  19578002.0           1444                    0.000074
5  West South Central  41685250.0           1961                    0.000047
6            Mountain  25514320.0           1096                    0.000043
7      South Atlantic  66781137.0           1714                    0.000026
8             Pacific  53229044.0           1217                    0.000023

In [33]:
df_articles_per_capita_region.sort_values(by='population', axis=0, ascending=False, ignore_index=True)

regional_division  population  article_title  articles_per_capita_region
0      South Atlantic  66781137.0           1714                    0.000026
1             Pacific  53229044.0           1217                    0.000023
2  East North Central  47097779.0           4491                    0.000095
3     Middle Atlantic  41910858.0           3538                    0.000084
4  West South Central  41685250.0           1961                    0.000047
5            Mountain  25514320.0           1096                    0.000043
6  West North Central  19721893.0           3382                    0.000171
7  East South Central  19578002.0           1444                    0.000074
8         New England  11503343.0           1314                    0.000114

**6.	Census divisions by high quality coverage: Rank ordered list of US census divisions (in descending order) by high quality articles per capita.**

In [34]:
# Source: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html
table_6 = df_high_quality_region.sort_values(by='high_quality_per_capita', axis=0, ascending=False, ignore_index=True)
table_6

regional_division  population  is_high_quality  high_quality_per_capita
0         New England  11503343.0                4             3.477250e-07
1     Middle Atlantic  41910858.0               11             2.624618e-07
2  East North Central  47097779.0               10             2.123242e-07
3      South Atlantic  66781137.0                7             1.048200e-07
4  East South Central  19578002.0                1             5.107774e-08
5            Mountain  25514320.0                1             3.919368e-08
6  West South Central  41685250.0                1             2.398930e-08
7             Pacific  53229044.0                1             1.878674e-08
8  West North Central  19721893.0                0             0.000000e+00

# References:

- [1]. Homework 2 - Considering Bias in Data.docx